webapp/src/main/java/com/box/l10n/mojito/service/delta/DeltaService.java

```diff
@@ -0,0 +1,192 @@
+            locales = repositoryService.getRepositoryLocalesWithoutRootLocale(repository)
+                    .stream().map(RepositoryLocale::getLocale).collect(Collectors.toList());
+        }
+
+        if (toDate == null) {
+            toDate = DateTime.now();
+        }
+
+        List<TMTextUnitVariant> variants = tmTextUnitVariantRepository.findAllUsedForRepositoryAndLocalesInDateRange(
+                        repository,
+                        locales,
+                        fromDate,
+                        toDate,
+                        pageable)
+                .getContent();
+
+        List<TextUnitVariantDelta> deltaVariants = variants.stream()
+                .map(this::getTextUnitVariantDelta)
+                .collect(Collectors.toList());
+
+        Map<String, DeltaLocaleDataDTO> deltaLocaleDataByBcp47Tags = getStringDeltaLocaleDataDTOMap(
+                deltaVariants);
+
+        DeltaMetadataDTO deltaMetadataDTO = new DeltaMetadataDTO();
+        deltaMetadataDTO.setFromDate(fromDate);
+        deltaMetadataDTO.setToDate(toDate);
+
+        DeltaResponseDTO deltaResponseDTO = new DeltaResponseDTO();
+        deltaResponseDTO.setTranslationsPerLocale(deltaLocaleDataByBcp47Tags);
+        deltaResponseDTO.setDeltaMetadataDTO(deltaMetadataDTO);
+
+        return deltaResponseDTO;
+    }
+
+    /**
+     * Queries the database to identify all Text Unit Variants that are newer and different than the variants that were
+     * used for the latest PullRuns.
+     * <p>
+     * To note: for the results to be accurate, each PushRun ID provided should have been processed before it's
+     * equivalent PullRun ID.
+     * <p>
+     * Overall logic looks like this:
+     *      1. Using the push_run_ids, collect a list of text units that were associated with the PushRuns
+     *      2. Using the pull_run_ids, create a list of the “base” translations that were included for
+     *      those PullRuns and collect a list of relevant text unit variants that correspond to the text units
+     *      3. Compare the “base” text unit variants with the “latest” variants and create a list of results with locale
+     *      information as well as the type of “translation change” that is being represented in the output
+     *      (e.g.: “NEW_TRANSLATION” vs. “UPDATED_TRANSLATION”).
+     *      4. Filter out any variants that were contained in the previous PullRuns or that have the same contents
+     *      as before.
+     *
+     * @return The delta of text unit variants, their translations and corresponding metadata.
+     */
+    public DeltaResponseDTO getDeltasForRuns(Repository repository,
+                                             List<Locale> locales,
+                                             List<PushRun> pushRuns,
+                                             List<PullRun> pullRuns) {
+
+        List<Long> localeIds = locales == null ? Collections.emptyList() :
```

Comment:
Optional.ofNullable(..) to replace the null check in the following lines?

Reply:
Not sure if I understood this, do you mean doing something like?
`        List<Long> localeIds = Optional.ofNullable(locales)
                        .map(l -> locales.stream().map(BaseEntity::getId).collect(Collectors.toList()))
                        .orElse(Collections.emptyList());`

That seems quite a bit more verbose than the current implementation.

Reply:
Similar. I'd start with something like that (which could be put into a static helper actually since it is doing it 3 times in a row) to get a stream from a nullable list

```
Optional.ofNullable(locales).map(List::stream).orElseGet(Stream::empty);
```

and have the specific code

Reply:
I've refactor the common stream code/ID parsing now.